### PR TITLE
Queryhandler concurrent access fix

### PR DESCRIPTION
**Pull request by theferrit32**
- add `python3-dev` header package to make psycopg2 happy
- construct QueryHandler in a `@app.on_event("startup")` function
- add a threading.Lock to query_handler, and acquire+release it when used

Since there's only one query_handler object, when fastapi schedules concurrent requests to the same worker thread, they will block when the call into `query_handler` happens if another concurrent request in the same thread is in that critical section.

EDIT: 
- no longer using `@app.on_event("startup")`
- use `queue.Queue` to handle locking and have a pool of more than 1 object

### CODE REVIEW

File: variation/main.py
Comment (korikuzma) [inline variation/main.py:369]:
```suggestion
```

**Review by korikuzma — APPROVED** [review]
Looks good to me! Just remove the print before you merge

**Review by jsstevenson — APPROVED** [review]
👍 if it works, it works

in the future, it'd be nice to put the lock in a place that's more proximate to the actual seqrepo calls so that the rest of the query method isn't blocking, but that's probably more trouble than it's worth right now

**Comment by theferrit32** [comment]
The exception below occurs *sometimes* when running uvicorn with the option `--workers 2` (or any number more than 1, the default). There was another exception too but I unfortunately didn't capture the log. That one was on a `remove()` call rather than an `open()` call. When that module is initialized when imported by one worker process, it writes/deletes some files in the same directory that the other workers are and sometimes this causes a problem. Depends on nondeterministic timing of those function calls across processes.

I've set `--workers` back to 1 in our deployment for now. It would be nice if it could run with more workers, but is not necessary. I've used a file-based Lock implementation before to do synchronization across processes in python, but there may be other better strategies.

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/uvicorn/_subprocess.py", line 76, in subprocess_started
    target(sockets=sockets)
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/uvicorn/server.py", line 59, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/uvicorn/server.py", line 66, in serve
    config.load()
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/uvicorn/config.py", line 471, in load
    self.loaded_app = import_from_string(self.app)
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/app/variation/main.py", line 21, in <module>
    from variation.query import QueryHandler
  File "/app/variation/query.py", line 8, in <module>
    from cool_seq_tool import SEQREPO_DATA_PATH, TRANSCRIPT_MAPPINGS_PATH, \
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/cool_seq_tool/__init__.py", line 29, in <module>
    d = DataDownload()
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/cool_seq_tool/data/data_downloads.py", line 19, in __init__
    self._download_data()
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/cool_seq_tool/data/data_downloads.py", line 30, in _download_data
    self._download_mane_summary(ftp)
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/cool_seq_tool/data/data_downloads.py", line 51, in _download_mane_summary
    with gzip.open(mane_data_path, "rb") as f_in:
  File "/usr/lib/python3.9/gzip.py", line 58, in open
    binary_file = GzipFile(filename, gz_mode, compresslevel)
  File "/usr/lib/python3.9/gzip.py", line 173, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
FileNotFoundError: [Errno 2] No such file or directory: '/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.9/site-packages/cool_seq_tool/data/MANE.GRCh38.v1.0.summary.txt.gz'
```

**Comment by jsstevenson** [comment]
Would swapping in Gunicorn for Uvicorn make this any simpler?

**Comment by theferrit32** [comment]
@jsstevenson I think gunicorn probably works in the same way, starting each worker process at the same time, and each of those will execute their initialization code including imports roughly at the same time. If so, then that is unlikely to fix the issue with multiple workers.

**Comment by theferrit32** [comment]
Merged `0.7.0-dev7` into this branch

**Comment by theferrit32** [comment]
@korikuzma I am seeing a thing where the process memory seems to monotonically grow over time. I'm going to see if just telling uvicorn to restart the worker process every 1000 requests or something will help that. One container that's been running for 18 days is at 1GB of RAM utilization.

Some requests are also hanging in non-deterministic ways. I'm running 5 containers and the same request will be completed by one container but hang in another container. Not sure the cause. Not sure if the async code in some of these functions is causing the lock to not be released properly all the time.

**Comment by theferrit32** [comment]
Adding this to the front of the Docker CMD makes it so I can run multiple workers:
`python -c "import cool_seq_tool" &&`

Example: 
`CMD python -c "import cool_seq_tool" && uvicorn variation.main:app --workers 4 --port 80 --host 0.0.0.0`

I also swapped my Dockerfile to no longer use pipenv because I was getting annoyed with pipenv performance, but I won't push those changes here.

**Comment by theferrit32** [comment]
@korikuzma @jsstevenson I think this is ready for merging into staging if there is approval from your side. I don't have write access here so someone else will need to merge.

I recently switched to using a different method of running multiple worker processes. That's here if you're interested: https://github.com/clingen-data-model/architecture/pull/565/files#diff-15e1c72eeec219b521bdc0d2ef8edfe21fe2a03fdb1343308c40c71d9bb2f10fR29

**Review by korikuzma — CHANGES_REQUESTED** [review]
Everything works. There are some flake8 warnings to resolve though:
```
./variation/main.py:64:1: E302 expected 2 blank lines, found 1
./variation/main.py:65:1: D103 Missing docstring in public function
./variation/main.py:65:27: ANN201 Missing return type annotation for public function
./variation/main.py:132:54: E128 continuation line under-indented for visual indent
./variation/main.py:438:89: E501 line too long (95 > 88 characters)
```

**Comment by korikuzma** [comment]
@theferrit32 said we can close this. He will be working on resolving the issue in seqrepo

**Pull request closed without merging**
